include/random_walks/nuts_hmc_walk.hpp

```diff
@@ -0,0 +1,384 @@
+    NT average_acceptance = 0;
+
+    // References to xs
+    Point x, v;
+
+    // Helper points
+    Point v_pl, v_min, v_min_j, v_pl_j, X_pl, X_pl_j, X_min, X, X_rnd_j, X_min_j, x_pl_min;
+
+    // Gradient function
+    NegativeGradientFunctor &F;
+
+    bool accepted;
+
+    // Burnin parameters
+    NT eps_step, mu, log_tilde_eps, H_tilde, alpha, na;
+    const NT delta = NT(0.65), Delta_max = NT(1000), gamma = NT(0.05), t0 = NT(10), kk = NT(0.85);
+
+    // Density exponent
+    NegativeLogprobFunctor &f;
+
+    Walk(Polytope *P,
+         Point &p,
+         NegativeGradientFunctor &neg_grad_f,
+         NegativeLogprobFunctor &neg_logprob_f,
+         parameters<NT, NegativeGradientFunctor> &param,
+         bool burn_in_phase = true) :
+         params(param),
+         F(neg_grad_f),
+         f(neg_logprob_f)
+    {
+      dim = p.dimension();
+
+      v_pl.set_dimension(dim);
+      v_min.set_dimension(dim);
+      v_min_j.set_dimension(dim);
+      v_pl_j.set_dimension(dim);
+      X_pl.set_dimension(dim);
+      X_pl_j.set_dimension(dim);
+      X_min.set_dimension(dim);
+      X.set_dimension(dim);
+      X_rnd_j.set_dimension(dim);
+      X_min_j.set_dimension(dim);
+      x_pl_min.set_dimension(dim);
+
+      eps_step = params.eta;
+      mu = std::log(10*eps_step);
+      log_tilde_eps = NT(0);
+      H_tilde = NT(0);
+      alpha = NT(0);
+      na = NT(0);
+
+      // Starting point is provided from outside
+      x = p;
+
+      accepted = false;
+
+      // Initialize solver
+      solver = new Solver(0, params.eta, pts{x, x}, F, bounds{P, NULL});
+      disable_adaptive();
+
+      if (burn_in_phase)
+      {
+        RandomNumberGenerator rng(dim);
+        burnin(rng);
+      }
+    };
+
+
+    inline void burnin(RandomNumberGenerator &rng,
+                       unsigned int N = 1000,
+                       unsigned int walk_length=1)
+    {
+      reset_num_runs();
+      Point p = x;
+      NT L;
+
+      if ((solver->get_bounds())[0] == NULL) 
+      {
+        L = (NT(100) / NT(dim)) * (NT(100) / NT(dim));
+      }
+      else
+      {
+        Polytope K = *(solver->get_bounds())[0];
+        L = estimate_L_smooth(K, p, walk_length, F, rng);
+      }
+
+      eps_step = NT(5) / (NT(dim) * std::sqrt(L));
+      solver->set_eta(eps_step);
+
+      for (int i = 0; i < N; i++)
+      {
+        apply(rng, walk_length, true);
+        solver->set_eta(eps_step);
+      }
+      reset_num_runs();
+    }
+
+
+    inline void apply(RandomNumberGenerator &rng,
+                      unsigned int walk_length=1,
+                      bool burnin = false)
+    {
+      num_runs++;
+
+      int x_counting_total = 0;
+
+      // Pick a random velocity
+      v = GetDirection<Point>::apply(dim, rng, false);
+      
+      v_pl = v;
+      v_min = NT(-1) * v;
+      X_pl = x;
+      X_min = x;
+      
+      NT h1 = hamiltonian(x, v);
+
+      NT uu = std::log(rng.sample_urdist()) - h1;
+      int j = -1;
+      bool s = true;
```

Comment:
what is `s`, does it make sense to give a more descriptive name ?

Reply:
this comes from the paper